libs/librepcb/common/systeminfo.cpp

```diff
@@ -279,6 +284,13 @@ QString SystemInfo::getProcessNameByPid(qint64 pid) {
   if (processName.endsWith(" (deleted)"))
     processName.chop(strlen(" (deleted)"));
 #elif defined(Q_OS_WIN32) || defined(Q_OS_WIN64)  // Windows
+  if (!isProcessRunning(pid)) {
```

Comment:
I understand that this might fix failed unit tests in some cases. But to be honest, somehow I don't like to call `isProcessRunning()` here since beside a unit test, `getProcessNameByPid()` is never called with a PID of a killed process. `DirectoryLock` is the only location where `getProcessNameByPid()` is called, and only with a PID previously checked with `isProcessRunning()`. Now with this change, the `isProcessRunning()` function is called two times every time we want to determine a process name.

IMHO we could simply change the documentation of `getProcessNameByPid()` that the returned value is not guaranteed to be correct if the PID does not represent a running process (i.e. the user has to call `isProcessRunning()` first). And then just remove the unit tests which call `getProcessNameByPid()` with a PID of a non-running process.

Or what do you think about this @andyfrommel?

Reply:
A failed unit test means either the production code or the unit test is faulty.
My first opinion was the production code needs a change - because the unit tests
seemed to work on all other users dev computer and the failure did only appear on
mine - but i was wrong. The unit test does not have a try-catch section, so it
does not consider all possible reactions of the tested method.

Removing a unit test that sometimes fails? better not, been there (even if the method
is not used anywhere else in the code).
A unit test should test the production code against the spec. It can happen that it
is not specified how a method should react under a certain condition and then its
result is not predictable for the unit test. In those cases all possible results should
be merged together as a combined result (or even better: the spec should be reworked)

I have reverted the code in systeminfo.cpp and changed the unit test accordingly.
The other two throw() cases in the tested method are still hard..impossible to
provoke in order to test them in a unit test.

Reply:
OK I agree it's better to adjust the test to also accept throwing exceptions in addition to returning an empty string :+1:

However, honestly I don't like some things in the current test implementation:

- It checks for the string "QueryFullProcessImageNameW() failed" - but this string is Windows specific so the test will be considered as failed if an exception is thrown on another operating system
- Since usually an empty string is returned (instead of throwing an exception), the "catch" part of the test is almost never executed. So if we ever change some implementation detail of the `getProcessNameByPid()` method which leads to a different exception message, the test will still be successful on most systems, but might fail on systems where an exception is thrown (like yours) - so we'll have the same problem with an unreliable unit test again :wink:
- The test is quite complex and verbose :)

What about a much simpler solution, which doesn't check the exception message (since it is pretty irrelevant)? Suggestion:

```cpp
try {
  processName = SystemInfo::getProcessNameByPid(pid);
} catch (const Exception&) {
  processName = QString();  // consider as process not running
}
EXPECT_EQ(QString(), processName) << qPrintable(processName);
```

Btw the coding style of the changes does not follow our coding style, see [here](https://developers.librepcb.org/df/d24/doc_code_style_guide.html) what coding style we have and how to format it automatically with clang-format. If you're not able to run the formatter, no problem, we could do it for you :slightly_smiling_face: